### ISSUE TITLE
Samsung doesn't like response to BrowseDirectChild

### ISSUE THREAD

**Issue by anacrolix** [anacrolix/dms#19]
When the samsung tv does a BrowseDirectChild, the response from dms isn't accepted.  Rygel works.

samsung to rygel

``` xml
POST /Control/MediaExport/RygelContentDirectory HTTP/1.0
HOST: 192.168.1.38:58635
CONTENT-LENGTH: 415
CONTENT-TYPE: text/xml;charset="utf-8"
USER-AGENT: DLNADOC/1.50
SOAPACTION: "urn:schemas-upnp-org:service:ContentDirectory:1#Browse"

<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle=
"http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:Browse xmlns:u="urn:schem
as-upnp-org:service:ContentDirectory:1"><ObjectID>0</ObjectID><BrowseFlag>Browse
DirectChildren</BrowseFlag><Filter>*</Filter><StartingIndex>0</StartingIndex><Re
questedCount>1</RequestedCount><SortCriteria></SortCriteria></u:Browse></s:Body>
</s:Envelope>
```

rygel to samsung

``` xml
<?xml version="1.0"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelo
pe/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:Brow
seResponse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1"><Result>&lt
;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quo
t;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:dlna=&quot;urn:schemas-dln
a-org:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lit
e/&quot;&gt;&lt;container parentID=&quot;0&quot; id=&quot;Filesystem&quot; child
Count=&quot;2&quot; searchable=&quot;1&quot; restricted=&quot;1&quot;&gt;&lt;dc:
title&gt;Files &amp;amp; Folders&lt;/dc:title&gt;&lt;upnp:class&gt;object.contai
ner.storageFolder&lt;/upnp:class&gt;&lt;upnp:storageUsed&gt;-1&lt;/upnp:storageU
sed&gt;&lt;upnp:containerUpdateID&gt;925&lt;/upnp:containerUpdateID&gt;&lt;upnp:
objectUpdateID&gt;925&lt;/upnp:objectUpdateID&gt;&lt;upnp:totalDeletedChildCount
&gt;2&lt;/upnp:totalDeletedChildCount&gt;&lt;upnp:searchClass includeDerived=&qu
ot;1&quot;&gt;object.item.imageItem&lt;/upnp:searchClass&gt;&lt;upnp:searchClass
 includeDerived=&quot;1&quot;&gt;object.item.imageItem.photo&lt;/upnp:searchClas
s&gt;&lt;upnp:searchClass includeDerived=&quot;1&quot;&gt;object.item.videoItem&
lt;/upnp:searchClass&gt;&lt;upnp:searchClass includeDerived=&quot;1&quot;&gt;obj
ect.item.audioItem&lt;/upnp:searchClass&gt;&lt;upnp:searchClass includeDerived=&
quot;1&quot;&gt;object.item.audioItem.musicTrack&lt;/upnp:searchClass&gt;&lt;upn
p:searchClass includeDerived=&quot;1&quot;&gt;object.item.playlistItem&lt;/upnp:
searchClass&gt;&lt;upnp:searchClass includeDerived=&quot;1&quot;&gt;object.conta
iner&lt;/upnp:searchClass&gt;&lt;res protocolInfo=&quot;http-get:*:text/xml:DLNA
.ORG_PN=DIDL_S;DLNA.ORG_FLAGS=00f00000000000000000000000000000&quot;&gt;http://1
92.168.1.38:58635/MediaExport/i/RmlsZXN5c3RlbQ&#37;3D&#37;3D/pl/DIDL_S.xml&lt;/res&gt;&l
t;/container&gt;&lt;/DIDL-Lite&gt;</Result><NumberReturned>1</NumberReturned><To
talMatches>3</TotalMatches><UpdateID>952</UpdateID></u:BrowseResponse></s:Body><
/s:Envelope>
```

samsung to dms

```
POST /ctl/ContentDirectory HTTP/1.0
HOST: 192.168.1.38:1338
CONTENT-LENGTH: 415
CONTENT-TYPE: text/xml;charset="utf-8"
USER-AGENT: DLNADOC/1.50
SOAPACTION: "urn:schemas-upnp-org:service:ContentDirectory:1#Browse"

<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle=
"http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:Browse xmlns:u="urn:schem
as-upnp-org:service:ContentDirectory:1"><ObjectID>0</ObjectID><BrowseFlag>Browse
DirectChildren</BrowseFlag><Filter>*</Filter><StartingIndex>0</StartingIndex><Re
questedCount>1</RequestedCount><SortCriteria></SortCriteria></u:Browse></s:Body>
</s:Envelope>
```

dms to samsung

```
<?xml version="1.0" encoding="UTF-8"?>
<Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/" encodingStyle="http:
//schemas.xmlsoap.org/soap/encoding/">
  <Body xmlns="http://schemas.xmlsoap.org/soap/envelope/"><BrowseResponse xmlns=
"urn:schemas-upnp-org:service:ContentDirectory:1">
  <TotalMatches>421</TotalMatches>
  <NumberReturned>1</NumberReturned>
  <Result>&lt;DIDL-Lite xmlns:dc=&#34;http://purl.org/dc/elements/1.1/&#34; xmln
s:upnp=&#34;urn:schemas-upnp-org:metadata-1-0/upnp/&#34; xmlns=&#34;urn:schemas-
upnp-org:metadata-1-0/DIDL-Lite/&#34; xmlns:dlna=&#34;urn:schemas-dlna-org:metad
ata-1-0/&#34;&gt;&lt;item id=&#34;1/cUA(!&amp;lt;EIE@VTIaF@ek$s+14O!W`F#@&amp;lt
;?/`$ig;7&amp;lt;,uDbF(9-!AH;i4!!#RiJcS4lEcl7QD.R6l@4u/&amp;#34;G&#37;E=$Gp#plF)u;=E
c`s8Bk1dnF&amp;#34;]7&#37;+B`8aEcW@AAft&amp;amp;dDf&amp;amp;p(+=84#EbT*&amp;amp;A90A
?E&amp;amp;7O&#34; parentID=&#34;0&#34; restricted=&#34;1&#34;&gt;&#xA;  &lt;upn
p:class&gt;object.item.videoItem&lt;/upnp:class&gt;&#xA;  &lt;dc:title&gt;10 Yea
rs of Thomas &amp;amp; Friends.mp4&lt;/dc:title&gt;&#xA;  &lt;res protocolInfo=&
#34;http-get:*:video/mp4:DLNA.ORG_OP=01;DLNA.ORG_CI=0&#34; size=&#34;949747293&#
34; bitrate=&#34;1954620&#34; duration=&#34;1:04:47.189333&#34; resolution=&#34;
696x480&#34;&gt;http://192.168.1.38:1338/res?path=&#37;2Frun&#37;2Fmedia&#37;2Fdave&#37;2FMy+Pas
sport&#37;2FVideos&#37;2F10+Years+of+Thomas+&#37;26+Friends.mp4&lt;/res&gt;&#xA;  &lt;res pr
otocolInfo=&#34;http-get:*:video/mpeg:DLNA.ORG_OP=10;DLNA.ORG_CI=1;DLNA_PN=MPEG_
PS_PAL&#34; resolution=&#34;720x576&#34;&gt;http://192.168.1.38:1338/res?path=&#37;2
Frun&#37;2Fmedia&#37;2Fdave&#37;2FMy+Passport&#37;2FVideos&#37;2F10+Years+of+Thomas+&#37;26+Friends.mp4&
amp;amp;transcode=t&lt;/res&gt;&#xA;&lt;/item&gt;&lt;/DIDL-Lite&gt;</Result>
  <UpdateID>1379701483</UpdateID>
</BrowseResponse></Body>
</Envelope>
```

---
- Bitbucket: https://bitbucket.org/anacrolix/dms/issue/19
- Originally reported by: David Mackay
- Originally created at: 2013-09-20T21:56:23.187

**Comment by anacrolix** [comment]
_From Matt Joiner on 2013-09-21 05:38:53+00:00_

David could you run this with the latest version of dms?

```
go get -u -v bitbucket.org/anacrolix/dms
```

I notice several things that are concerning with the samples you gave:
- rygel and dms are sharing different filesystem locations? Could you test sharing the same location?
- dms in the test exhibiting the parentID bug which was recently fixed. It's also using the nasty gob-encoded IDs which make debugging harder. This has also been fixed in the tip.

Can you comment on how this issue relates to issue #4? Is the connection disconnect described in issue #4 correctly resolved, does the behaviour in this bug now occur as a result of that fix? (That is to say, it's getting further than it used to do, before I put in the fix for issue #4?).

Thanks a lot for your aid in this.

**Comment by anacrolix** [comment]
_From David Mackay on 2013-09-21 11:55:25+00:00_

In the posting, rygel and dms are using the same file location.  In the libpcap dumps that I sent earlier, they weren't.

Issue #4 is resolved.  That gets us to this issue.  I'll pull the latest mercurial and retest.

**Comment by anacrolix** [comment]
_From David Mackay on 2013-10-11 19:09:18+00:00_

Sorry, I haven't dropped off the face of the earth.  The circa 2002 pre-pentium amd system that I used for various things, including email, did.  I'm back now.  I've taken another look, and there may be a design level problem, at least with a setup like mine.  I'm running dms on a laptop with a 2 TB usb drive on it.  The Samsung TV appears to have a 20 second timeout, and it appears as though readContainer doesn't always complete within the timeout period.  I also see that you're only calling readContainer on a browse request.  I don't know if MS Windows has the equivalent of inotify, but it seems to me that calling readContainer on startup, and then using inotify to pick up any file system changes would be a good way to go.

However, when dms does get in a response, the Samsung still doesn't like it.  Below are the results from rygel, and a copy of dms that I cloned this morning.

dms started with:
./main -friendlyName=dms -ifname=em1 -path=/

dms response to BrowseDirectChildren

```
#!xml
<?xml version="1.0" encoding="UTF-8"?>
<Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/" encodingStyle="http:
//schemas.xmlsoap.org/soap/encoding/">
  <Body xmlns="http://schemas.xmlsoap.org/soap/envelope/"><BrowseResponse xmlns=
"urn:schemas-upnp-org:service:ContentDirectory:1">
  <TotalMatches>20</TotalMatches>
  <NumberReturned>1</NumberReturned>
  <Result>&lt;DIDL-Lite xmlns:dc=&#34;http://purl.org/dc/elements/1.1/&#34; xmln
s:upnp=&#34;urn:schemas-upnp-org:metadata-1-0/upnp/&#34; xmlns=&#34;urn:schemas-
upnp-org:metadata-1-0/DIDL-Lite/&#34; xmlns:dlna=&#34;urn:schemas-dlna-org:metad
ata-1-0/&#34;&gt;&lt;container id=&#34;1/bin&#34; parentID=&#34;0&#34; restricte
d=&#34;1&#34; childCount=&#34;2386&#34;&gt;&#xA;  &lt;upnp:class&gt;object.conta
iner.storageFolder&lt;/upnp:class&gt;&#xA;  &lt;dc:title&gt;bin&lt;/dc:title&gt;
&#xA;&lt;/container&gt;&lt;/DIDL-Lite&gt;</Result>
  <UpdateID>1381511812</UpdateID>
</BrowseResponse></Body>
</Envelope>
```

The Samsung TV gives me a popup box stating that there are no media files in this folder

Rygel response to BrowseDirectChildren:

```
#!xml
<?xml version="1.0"?><s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelo
pe/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><s:Body><u:Brow
seResponse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1"><Result>&lt
;DIDL-Lite xmlns:dc=&quot;http://purl.org/dc/elements/1.1/&quot; xmlns:upnp=&quo
t;urn:schemas-upnp-org:metadata-1-0/upnp/&quot; xmlns:dlna=&quot;urn:schemas-dln
a-org:metadata-1-0/&quot; xmlns=&quot;urn:schemas-upnp-org:metadata-1-0/DIDL-Lit
e/&quot;&gt;&lt;container parentID=&quot;0&quot; id=&quot;Filesystem&quot; child
Count=&quot;2&quot; searchable=&quot;1&quot; restricted=&quot;1&quot;&gt;&lt;dc:
title&gt;Files &amp;amp; Folders&lt;/dc:title&gt;&lt;upnp:class&gt;object.contai
ner.storageFolder&lt;/upnp:class&gt;&lt;upnp:storageUsed&gt;-1&lt;/upnp:storageU
sed&gt;&lt;upnp:containerUpdateID&gt;925&lt;/upnp:containerUpdateID&gt;&lt;upnp:
objectUpdateID&gt;925&lt;/upnp:objectUpdateID&gt;&lt;upnp:totalDeletedChildCount
&gt;2&lt;/upnp:totalDeletedChildCount&gt;&lt;upnp:searchClass includeDerived=&qu
ot;1&quot;&gt;object.item.imageItem&lt;/upnp:searchClass&gt;&lt;upnp:searchClass
 includeDerived=&quot;1&quot;&gt;object.item.imageItem.photo&lt;/upnp:searchClas
s&gt;&lt;upnp:searchClass includeDerived=&quot;1&quot;&gt;object.item.videoItem&
lt;/upnp:searchClass&gt;&lt;upnp:searchClass includeDerived=&quot;1&quot;&gt;obj
ect.item.audioItem&lt;/upnp:searchClass&gt;&lt;upnp:searchClass includeDerived=&
quot;1&quot;&gt;object.item.audioItem.musicTrack&lt;/upnp:searchClass&gt;&lt;upn
p:searchClass includeDerived=&quot;1&quot;&gt;object.item.playlistItem&lt;/upnp:
searchClass&gt;&lt;upnp:searchClass includeDerived=&quot;1&quot;&gt;object.conta
iner&lt;/upnp:searchClass&gt;&lt;res protocolInfo=&quot;http-get:*:text/xml:DLNA
.ORG_PN=DIDL_S;DLNA.ORG_FLAGS=00f00000000000000000000000000000&quot;&gt;http://1
92.168.1.38:58635/MediaExport/i/RmlsZXN5c3RlbQ%3D%3D/pl/DIDL_S.xml&lt;/res&gt;&l
t;/container&gt;&lt;/DIDL-Lite&gt;</Result><NumberReturned>1</NumberReturned><To
talMatches>3</TotalMatches><UpdateID>952</UpdateID></u:BrowseResponse></s:Body><
/s:Envelope>


```

**Comment by anacrolix** [comment]
_From [mkorenkov](https://bitbucket.org/mkorenkov) on 2014-01-04 15:24:24+00:00_

Guys, is there any way I can get DLNA specification to help you with this?

**Comment by anacrolix** [comment]
_From Matt Joiner on 2014-01-04 16:16:09+00:00_

I believe it's a shortcoming in the Go XML packages. `encoding/xml` doesn't properly handle namespaces when marshalling. The Samsung TV is likely using a particularly pedantic/correct XML parser and not finding the required elements due to their incorrect namespacing. Issue #25 is related. Also see Go's https://code.google.com/p/go/issues/detail?id=6800.

**Comment by anacrolix** [comment]
_From Matt Joiner on 2014-01-13 16:33:20+00:00_

This should be fixed by fbd23ce. Please give feedback! Thanks.
